File: 739. Daily Temperatures/DailyTemperatures.py
'''
https://leetcode.com/problems/daily-temperatures/
'''


class Solution:
    def dailyTemperatures(self, temperatures: list[int]) -> list[int]:
        n = len(temperatures)
        stack = []
        answer = [0] * n
        
        for i in range(n):
            temp = temperatures[i]
            while stack and temp > temperatures[stack[-1]]:
                answer[stack[-1]] = i - stack[-1]
                stack.pop()
            
            stack.append(i)
            
        return answer

File: 739. Daily Temperatures/DailyTemperatures.java
import java.util.Arrays;
import java.util.Stack;

public class DailyTemperatures {
    public int[] dailyTemperatures(int[] temperatures) {
        int[] answer = new int[temperatures.length];
        Stack<Integer> lowerTemps = new Stack<>();

        int i = 0;
        for (int temp: temperatures){
            while (!lowerTemps.isEmpty() && temperatures[lowerTemps.peek()] < temp)
                answer[lowerTemps.peek()] = i - lowerTemps.pop();

            lowerTemps.push(i);
            i++;
        }

        return answer;
    }


    public static void main(String[] args) {
        DailyTemperatures dt = new DailyTemperatures();
        int[] temperatures = {30,40,50,60};
        System.out.println(Arrays.toString(dt.dailyTemperatures(temperatures)));
    }

}
